chain/bitcoind_events_test.go

```diff
@@ -0,0 +1,434 @@
+	}
+
+	ntfns := client.Notifications()
+
+	miner.Client.Generate(1)
+
+	// We cant be sure what order the notifications will arrive in, but
+	// we know that we expect three of them. We expect one of each of the
+	// following.
+	var (
+		relevantTx             int
+		filteredBlockConnected int
+		blockConnected         int
+	)
+	for i := 0; i < 3; i++ {
+		select {
+		case ntfn := <-ntfns:
+			switch ntfnType := ntfn.(type) {
+			case RelevantTx:
+				require.True(
+					t, ntfnType.TxRecord.Hash.IsEqual(
+						&hash,
+					),
+				)
+				relevantTx++
+				if relevantTx != 1 {
+					t.Fatalf("received more RelevantTx " +
+						"notifications than expected")
+				}
+
+			case FilteredBlockConnected:
+				filteredBlockConnected++
+				if filteredBlockConnected != 1 {
+					t.Fatalf("received more " +
+						"FilterBlockRequest " +
+						"notifications than expected")
+				}
+
+			case BlockConnected:
+				blockConnected++
+				if blockConnected != 1 {
+					t.Fatalf("received more " +
+						"BlockConnected " +
+						"notifications than expected")
+				}
+
+			default:
+				t.Fatalf("reveived unexpected %T "+
+					"notification", ntfn)
+			}
+
+		case <-time.After(time.Second):
+			t.Fatalf("timed out waiting for block notification")
+		}
+	}
+}
+
+// testReorg tests that the given BitcoindClient correctly responds to a chain
+// re-org.
+func testReorg(t *testing.T, miner1, miner2 *rpctest.Harness,
+	client *BitcoindClient) {
+
+	ntfns := client.Notifications()
+	miner1Hash, commonHeight, err := miner1.Client.GetBestBlock()
+	if err != nil {
+		t.Fatalf("could not fetch best block from miner1")
+	}
+
+	miner2Hash, miner2Height, err := miner2.Client.GetBestBlock()
+	if err != nil {
+		t.Fatalf("could not fetch best block from miner2")
+	}
+
+	require.Equal(t, commonHeight, miner2Height)
+	require.Equal(t, miner1Hash, miner2Hash)
+
+	// Let miner2 generate a few blocks and ensure that our bitcoind client
+	// is notified of this block.
+	hashes, err := miner2.Client.Generate(5)
+	if err != nil {
+		t.Fatalf("could not generate block with miner2: %v", err)
+	}
+	require.Len(t, hashes, 5)
+
+	for i := 0; i < 5; i++ {
+		commonHeight++
+		ntfnHeight, ntfnHash := waitForBlockNtfn(t, ntfns)
+		require.Equal(t, commonHeight, ntfnHeight)
+		require.True(t, ntfnHash.IsEqual(hashes[i]))
+	}
+
+	// Now disconnect the two miners.
+	err = miner1.Client.AddNode(miner2.P2PAddress(), rpcclient.ANRemove)
+	if err != nil {
+		t.Fatalf("unable to remove node: %v", err)
+	}
+
+	// Generate 5 blocks on miner2.
+	_, err = miner2.Client.Generate(5)
+	if err != nil {
+		t.Fatalf("could not generate block with miner2: %v", err)
+	}
+
+	// Since the miners have been disconnected, we expect not to get any
+	// notifications from our client since our client is connected to
+	// miner1.
+	select {
+	case ntfn := <-ntfns:
+		t.Fatalf("received a notification of type %T but expected,"+
+			" none", ntfn)
+	case <-time.After(time.Millisecond * 10):
```

Comment:
This seems quite short. Should we wait for a bit longer, just to be sure? Maybe half a second? Shouldn't add too much to the overall test duration, I think.